### PR TITLE
test: use MiniWallet for mempool_updatefromblock.py

### PR DESCRIPTION
**Pull request by pg156**
This PR enables a non-wallet functional test `mempool_updatefromblock.py` to be run without wallet compiled, by using the MiniWallet proposed in #20078. This picks up from where #21999 left off.

Commit 1 of 2: MiniWallet changes only to minimize diffs for easier reviewing
Commit 2 of 2: Refactor (renaming a function and variables)

The test run time decreases from 16.5s to 4.3s.

### CODE REVIEW

File: test/functional/mempool_updatefromblock.py [hunk 1]
```diff
@@ -52,73 +48,49 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, start_input_txid='', e
...
 
     def run_test(self):
         # Use batch size limited by DEFAULT_ANCESTOR_LIMIT = 25 to not fire "too many unconfirmed parents" error.
-        self.transaction_graph_test(size=100, n_tx_to_mine=[25, 50, 75])
+        self.transaction_list_test(size=100, n_tx_to_mine=[25, 50, 75])
 
 
 if __name__ == '__main__':
     MempoolUpdateFromBlockTest().main()
+
```
Comment (brunoerg) [inline test/functional/mempool_updatefromblock.py:96]: There is a whitespace here, you have to remove it to pass the linter.

Comment (pg156) [reply test/functional/mempool_updatefromblock.py]: Thanks. Fixed. Don't know why the linter now complains about the spelling errors in unchanged files.

^---- failure generated from test/lint/lint-python.sh
src/random.h:92: occuring ==> occurring
src/rpc/blockchain.cpp:794: nWe ==> new
src/util/syscall_sandbox.cpp:128: creat ==> create
test/functional/data/rpc_decodescript.json:81: ba ==> by, be
test/functional/data/rpc_decodescript.json:84: ba ==> by, be
^ Warning: codespell identified likely spelling errors. Any false positives? Add them to the list of ignored words in test/lint/lint-spelling.ignore-words.txt

Comment (brunoerg) [reply test/functional/mempool_updatefromblock.py]: Cool, now you should squash the commits. See: https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

Comment (pg156) [reply test/functional/mempool_updatefromblock.py]: Thanks. Done.

Comment (mzumsande) [reply test/functional/mempool_updatefromblock.py]: The problem of the linter now is "test/functional/mempool_updatefromblock.py:12:1: F401 'decimal.Decimal' imported but unused" - this PR removes all uses.

**Comment by MarcoFalke** [comment]
Are you still working on this?

**Comment by pg156** [comment]
> Are you still working on this?

Yes. Read mzumsande's comment too quickly and didn't realize it was a request to change. Will fix soon.

**Comment by Ferrydh** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by pg156** [comment]
The linter problem is fixed.

**Review by aureleoules — COMMENTED** [review]
tACK 203a3e065ecd0ec8a9fb84f4deb512ef2b66cdf7.
Tested on NixOS 22.05 64 bits with `--disable-wallet`.

I verified that the migration to MiniWallet does not alter what is being tested.
It takes 3s to run on my machine 🚀

File: test/functional/mempool_updatefromblock.py [hunk 2]
```diff
@@ -109,7 +81,7 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, start_input_txid='', e
         self.log.info('Checking descendants/ancestors properties of all of the in-mempool transactions...')
         for k, tx in enumerate(tx_id):
             self.log.debug('Check transaction #{}.'.format(k))
-            entry = self.nodes[0].getmempoolentry(tx)
+            entry = self.nodes[0].getrawmempool(True)[tx['txid']]
```
Comment (MarcoFalke) [inline test/functional/mempool_updatefromblock.py]: Why is this changed?

File: test/functional/mempool_updatefromblock.py [hunk 3]
```diff
@@ -52,42 +49,17 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, start_input_txid='', e
...
-                for _ in range(n_outputs):
-                    outputs[self.nodes[0].getnewaddress()] = output_value
-            else:
-                output_value = (inputs_value - fee).quantize(Decimal('0.00000001'))
-                outputs = {end_address: output_value}
+                parent_utxo = miniwallet.get_utxo(txid=tx_id[i-1]['txid'])
 
-            self.log.debug('output_value={}'.format(output_value))
-            self.log.debug('outputs={}'.format(outputs))
+            tx = miniwallet.send_self_transfer(from_node=self.nodes[0], utxo_to_spend=parent_utxo, sequence=SEQUENCE_FINAL)
```
Comment (MarcoFalke) [inline test/functional/mempool_updatefromblock.py]: Why does this need `SEQUENCE_FINAL`?

File: test/functional/mempool_updatefromblock.py [hunk 4]
```diff
@@ -52,42 +49,17 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, start_input_txid='', e
...
+            tx_id.append(tx)
 
-            # Create a new transaction.
-            unsigned_raw_tx = self.nodes[0].createrawtransaction(inputs, outputs)
-            signed_raw_tx = self.nodes[0].signrawtransactionwithwallet(unsigned_raw_tx)
-            tx_id.append(self.nodes[0].sendrawtransaction(signed_raw_tx['hex']))
-            tx_size.append(self.nodes[0].getmempoolentry(tx_id[-1])['vsize'])
+            vsize = self.nodes[0].getrawmempool(True)[tx['txid']]['vsize']
```
Comment (MarcoFalke) [inline test/functional/mempool_updatefromblock.py]: Why is this changed?

**Review by maflcko — APPROVED** [review]

File: test/functional/mempool_updatefromblock.py (hunk 2 above)
Comment (pg156) [reply test/functional/mempool_updatefromblock.py]: Changed back to `getmempoolentry`.

I chose `getrawmempool` over `getmempoolentry` to follow some existing functional test files. After a closer look at those two RPCs, I agree the existing `getmempoolentry` here is better as it indicates the "query using a transaction id" intention.

File: test/functional/mempool_updatefromblock.py (hunk 4 above)
Comment (pg156) [reply test/functional/mempool_updatefromblock.py]: Changed back to `getmempoolentry`.

File: test/functional/mempool_updatefromblock.py (hunk 3 above)
Comment (pg156) [reply test/functional/mempool_updatefromblock.py]: Without `SEQUENCE_FINAL`, the test broke at `assert_equal(len(node.getrawmempool()), size)` after the first block is invalidated, as the mempool has 26 transactions instead of the expected 100.

In a simpler example with fewer transactions without `SEQUENCE_FINAL`, I can see after the block is invalidated, transactions are removed from the mempool because of this place: https://github.com/bitcoin/bitcoin/blob/f05cf59d91eb03857dd9bdcc77607764da0349d2/src/consensus/tx_verify.cpp#L53

**Comment by MarcoFalke** [comment]
You'll need to do the changes from yesterday in the first commit, not in the second.

**Comment by pg156** [comment]
Fixed. Thank you @MarcoFalke.

File: test/functional/mempool_updatefromblock.py [hunk 5]
```diff
@@ -9,41 +9,36 @@
...
         Keyword arguments:
-        size -- the order N of the tournament which is equal to the number of the created transactions
+        size -- the number of the created transactions
         n_tx_to_mine -- the number of transaction that should be mined into a block
 
         If all of the N created transactions tx[0]..tx[N-1] reside in the mempool,
         the following holds:
             the tx[K] transaction:
             - has N-K descendants (including this one), and
             - has K+1 ancestors (including this one)
```
Comment (mjdietzx) [inline test/functional/mempool_updatefromblock.py:36]: I don't think these conditions hold true after this PR

**Review by mjdietzx — CHANGES_REQUESTED** [review]
I think this PR introduces an unwanted behavior change. It looks like now you have a simple straight line of spends tx0->tx1->...->txN.

Before this test created a huge, complicated "cyclic tournament" graph of transactions where:
- Transaction tx[K] is a child of each of previous transactions tx[0]..tx[K-1] at their output K-1
- Transaction tx[K] is an ancestor of each of subsequent transactions tx[K+1]..tx[N-1].

That's kind of hard to visualize. But the first transaction `tx0` has 99 outputs, and tx1, ... tx100 all spend one of tx0's outputs. And tx100 has 99 inputs, an input from tx1, ... tx99. Looking at L51-L80 (before this PRs changes) you can try to imagine what tx4 looks like and that's kinda hard!

It's easy to see by the runtime of this test. Before this PR running this test in our CI took ~5 minutes. Now this test runs in ~1s.

---

I've found this test to be extremely valuable bc it highlights a possible denial of service attack vector. Imagine someone creates a graph of transactions like this and gets them all mined into a single block (let's just say they mine the block or collude w/ a miner). Even though the mempool has `DEFAULT_ANCESTOR_LIMIT` and `DEFAULT_DESCENDANT_LIMIT` of 25, this test just shows how large and complicated a web of dependent transactions you can make within those limits.

Now imagine that huge graph takes up a single block that's mined. And then somehow (maybe on purpose) that block gets reorged. All of a sudden this entire block of transactions enters the mempool again. And all the nodes need to validate every single transaction in this graph before adding them back to their mempool. Imagine if we did something stupid (or a lot of extra looping work) in the mempool validation logic. When that block gets reorged it could potentially crash or stall a bunch of nodes.

So I think this test is a really good stress test as-is. If you do any looping or weird stuff in the mempool validation logic this test can very easily timeout. I learned this the hard way when trying out some RBF changes, where I saw this test timing out. And I started to really appreciate the value of this specific test at that point

**Comment by mjdietzx** [comment]
Concept ACK - I think this is a valuable test and it'll be good to update it so that it's not skipped when the wallet isn't compiled. My suggestion would be to minimize changes as much as possible to make it work w/ MiniWallet - and implementing this test as-is w/ MiniWallet will probably be difficult! But I think it'll be a good exercise, and I think MiniWallet can probably support what this test needs without any internal code changes to it (although I'm not totally sure bc I haven't looked at that code in a while)

**Comment by pg156** [comment]
Thank you so much @mjdietzx for the review with the context and insights. I am changing this to draft status, to explore how to keep the graph structure.
